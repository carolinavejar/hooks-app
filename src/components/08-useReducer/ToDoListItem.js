import React from 'react'

export const ToDoListItem = ({todo, i, handleDelete, handleToggle}) => {
    return (
        <li
            key = { todo.id }
            className = "list-group-item"
        >
            <small
                className = { `${ todo.done &&  'complete' }`}
                onClick = { () => {  handleToggle(todo.id) } }
            >
                    { i + 1 }. {todo.desc}
            </small>

            <button 
                className="btn btn-outline-primary ml-3 btnBorrar" 
                id = { todo.id }
                onClick = { handleDelete }
            >
                Borrar
            </button>
        </li>
    )
}
