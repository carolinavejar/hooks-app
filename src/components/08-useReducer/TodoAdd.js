import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAdd }) => {
    const [{ description }, handleInputChange, reset] = useForm({
        description: '',
    }); 

    const handleSubmit = (e) => {
        e.preventDefault();
        if(description.trim().length <= 0){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        handleAdd(newTodo);
        reset();
    }

    return (
        <>
            <h4>Agregar TO DO</h4>
            <hr />
            <form onSubmit = { handleSubmit }>
                <div className = "d-grid gap-2">
                    <input
                        className = "form-control"
                        type = "text"
                        name = "description"
                        placeholder = "Aprender ..."
                        autoComplete = "off"
                        value = { description }
                        onChange = { handleInputChange }
                    />
                    <button 
                        type = "submit"
                        className="btn btn-outline-primary btn-block mt-1" 
                    >
                        Agregar
                    </button>
                </div>
                
            </form>
        </>
    )
}
