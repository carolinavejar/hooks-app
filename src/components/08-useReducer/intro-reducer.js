const initialState = [ {
    id: 1,
    todo: 'Comprar pan', 
    done: false
}];

const todoReducer = (state = initialState, action) => {
    if(action?.type == "add") {
        return [...state, action.payload]
    }
    return state;
};

let toDos = todoReducer();
// aÃ±adir nuevo todo 
const newTodo = {
    id: 2,
    todo: 'Comprar chocolate', 
    done: false
};

const addToDoAction = {
    type: 'add',
    // newTodo
    payload : newTodo // ->estandar
}

toDos = todoReducer(toDos, addToDoAction);

console.log(toDos);