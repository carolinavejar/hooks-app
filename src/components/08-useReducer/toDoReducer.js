export const todoReducer = ( state = [], action ) => {
    switch (action.type) {
        case 'add':
            return [...state, action.payload];
    
        case 'remove': 
            // The filter() method creates a new array with all elements that pass the test implemented by the provided function
            return state.filter( state2 => state2.id != action.payload.id);

        case 'toggle' :
            return state.map(todo => 
               todo.id == action.payload ? { ...todo, done: !todo.done } : todo
            );
        case 'toggle-old': 
            return state.map(todo => {
                if( todo.id == action.payload) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                } else {
                    return todo;
                }
            })
        default:
            return state;
    }
}