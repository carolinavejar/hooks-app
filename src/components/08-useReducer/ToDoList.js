import React from 'react'
import { ToDoListItem } from './ToDoListItem';

export const ToDoList = ({todos, handleDelete, handleToggle}) => {
    console.log(todos);
    return (
        
        <ul>
            {
                todos.map( (todo, i) => (
                    <ToDoListItem
                        key = { todo.id }
                        todo = { todo }
                        i = { i }
                        handleDelete = { handleDelete }
                        handleToggle = { handleToggle}
                    />
                ))
            }
        </ul>
        
    )
}
