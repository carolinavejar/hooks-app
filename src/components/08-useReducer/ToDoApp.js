import React, { useReducer } from 'react';
import { useEffect } from 'react/cjs/react.development';
import '../css/index.css';
import { TodoAdd } from './TodoAdd';
import { ToDoList } from './ToDoList';
import { todoReducer } from './todoReducer';

export const TodoApp = () => {

    const init = () => {
        return JSON.parse(localStorage.getItem('todos')) || [];
    }

    const [ todos, dispatch ] = useReducer(todoReducer, [],  init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = ( id ) => {
        console.log( id );

        const deleteTodo = {
            id: id,
        }

        const action = {
            type: 'remove',
            payload: deleteTodo
        };

        dispatch(action);

    }

    const handleToggle = (pID) => {
        dispatch( {
            type: 'toggle',
            payload: pID
        });
    };

    const handleAdd = ( newTodo) =>{

        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    return (
        <div id = "todoList">
            <h1>To Do App ( {todos.length} )</h1>
            <hr />

            <div className = "row">
                <div className="col-7">
                    <ToDoList
                        todos = { todos }
                        handleDelete = { handleDelete}
                        handleToggle = { handleToggle }
                    />
                </div>
                <div className="col-5">
                    <TodoAdd
                        handleAdd = { handleAdd }
                    />
                </div>
            </div>
        </div>
    )
}
