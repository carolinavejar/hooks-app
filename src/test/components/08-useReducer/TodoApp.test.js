import React from 'react';
import { TodoApp }  from '../../../components/08-useReducer/TodoApp';
import '@testing-library/jest-dom';
import { demoTodos } from '../../fixtures/demoTodos';

import { shallow, mount } from 'enzyme/build';

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { createSerializer } from 'enzyme-to-json';
import { act } from 'react-dom/test-utils';

Enzyme.configure({ adapter: new Adapter() });
expect.addSnapshotSerializer( createSerializer ( { mode : 'deep' } ));

describe('Pruebas en el <ToDoApp /> ', () => {

    const wrapper = shallow(<TodoApp/>);

    Storage.prototype.setItem = jest.fn(()=> {});
    test('debe de mostrar <ToDoApp /> correctamente', () => {
        expect (wrapper).toMatchSnapshot();
    });

    test('debe agregar un ToDo', () => {
        const wrapper = mount(<TodoApp/>);

        act(()=> {
            wrapper.find('TodoAdd').prop('handleAdd')(demoTodos[0]);
            wrapper.find('TodoAdd').prop('handleAdd')(demoTodos[1]);
        })
        
        expect(wrapper.find('h1').text().trim()).toBe('To Do App ( 2 )');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });

    test('debe eliminar un ToDo', () => {
        wrapper.find('TodoAdd').prop('handleAdd')(demoTodos[0]);
        wrapper.find('ToDoList').prop('handleDelete')(demoTodos[0].id);
        expect(wrapper.find('h1').text().trim()).toBe('To Do App ( 0 )');
    });
})