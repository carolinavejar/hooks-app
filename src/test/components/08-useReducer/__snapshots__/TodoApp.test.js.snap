// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <ToDoApp />  debe de mostrar <ToDoApp /> correctamente 1`] = `
<div
  id="todoList"
>
  <h1>
    To Do App ( 
    0
     )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <ToDoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAdd={[Function]}
      />
    </div>
  </div>
</div>
`;
